src/pepr/operator/controllers/keycloak/authservice/authservice.ts

```diff
@@ -0,0 +1,120 @@
+import { Log, R } from "pepr";
+import { UDSConfig } from "../../../../config";
+import { Store } from "../../../common";
+import { UDSPackage } from "../../../crd";
+import { apiCall } from "../client-sync";
+import { Client } from "../types";
+import { updatePolicy } from "./authorization-policy";
+import { getAuthserviceConfig, operatorConfig, updateAuthServiceSecret } from "./config";
+import { Action, AuthServiceEvent, AuthserviceConfig, Chain } from "./types";
+
+export async function authservice(pkg: UDSPackage) {
+  // Get the list of clients from the package
+  const authServiceClients = R.filter(
+    sso => R.isNotNil(sso.enableAuthserviceSelector),
+    pkg.spec?.sso || [],
+  );
+
+  for (const client of authServiceClients) {
+    const name = `sso-client-${client.clientId}`;
+    const token = Store.getItem(name);
+    if (!token) {
+      throw new Error(`Failed to get token for client ${client.clientId}`);
+    }
+    const keycloakClient = await apiCall(client, "GET", token);
+
+    client.secret = keycloakClient.secret;
+    await reconcileAuthservice(
+      { name: client.clientId, action: Action.Add, client: client as unknown as Client },
+      client.enableAuthserviceSelector as { [key: string]: string },
+      pkg,
+    );
+  }
+
+  const authserviceClients = authServiceClients.map(client => client.clientId);
+
+  await purgeAuthserviceClients(pkg, authserviceClients);
+
+  return authserviceClients;
+}
+
+export async function purgeAuthserviceClients(pkg: UDSPackage, authserviceClients: string[] = []) {
+  // compute set difference of pkg.status.authserviceClients and authserviceClients using Ramda
+  R.difference(pkg.status?.authserviceClients || [], authserviceClients).forEach(async clientId => {
+    Log.info(`Removing stale authservice chain for client ${clientId}`);
+    await reconcileAuthservice({ name: clientId, action: Action.Remove }, {}, pkg);
+  });
+}
```

Comment:
Considering this function is called `purgeAuthserviceClients()` it kind of feels that like the expected behavior is to purge what you pass to it, so rename the argument to `newAuthserviceClients` for clarity.
```suggestion
export async function purgeAuthserviceClients(pkg: UDSPackage, newAuthserviceClients: string[] = []) {
  // compute set difference of existing pkg.status.authserviceClients and newAuthserviceClients using Ramda
  R.difference(pkg.status?.authserviceClients || [], newAuthserviceClients).forEach(async clientId => {
    Log.info(`Removing stale authservice chain for client ${clientId}`);
    await reconcileAuthservice({ name: clientId, action: Action.Remove }, {}, pkg);
  });
}
```

Reply:
Updated